### ISSUE TITLE
Electric Field application on periodic systems

### ISSUE THREAD

**Issue by Simonsays095** [pyscf/pyscf#586]
Hello there,

I'm trying to probe the response of a (1-dimensional) periodic system to an applied electric field. For now, I'm using a simple 1-dimensional Hydrogen chain as a model system, to make sure the calculation behaves as expected. Turns out, it isn't behaving as I expect.

I started with the example [here](https://github.com/pyscf/pyscf-doc/blob/master/examples/scf/40-apply_electric_field.py) for the molecular case, and extended it to periodic molecules like this:

```import numpy as np
from pyscf.pbc import gto, scf

cell = gto.Cell()

cell.atom = [
    ["H", ( 0, 0, 0)],
    ["H", ( 1, 0, 0)],
    ["H", ( 2, 0, 0)],
    ["H", ( 3, 0, 0)],
]
cell.a = np.eye(3)*4

cell.basis = '6-31g'
cell.dimension = 1
cell.unit = 'A'
cell.verbose = 6
cell.symmetry = 0
cell.charge = 0

cell.build()

EField_dir = np.asarray([1, 0, 0])
E_mags = np.linspace(-0.001, 0.001, 11, endpoint=True)

mean = cell.atom_coords().mean(axis=0)
cell.set_common_orig(mean)  # The gauge origin for dipole integral

class EFieldSCF(scf.hf.RHF):
    def get_hcore(self, cell=None, kpt=None):
        if cell is None: cell = self.cell
        if kpt is None: kpt = self.kpt
        return super().get_hcore(cell, kpt) + np.einsum('x,xij->ij', self.E, cell.pbc_intor('int1e_r', 3, 1, self.kpt))

Energy = []
for mag in E_mags:
    mf = EFieldSCF(cell).density_fit()
    mf.E = mag*EField_dir # mag*condDir

    mf.kernel() # gamma-point calculation
    Energy.append(mf.e_tot)

print("E_mag\t\tE_HF")
for data in zip(E_mags, Energy):
    print(*data, sep='\t\t')
```

The molecular case is a parabola, as I expect for small perturbations. However, the periodic case is both asymmetric and not a good fit to a polynomial:
![image](https://user-images.githubusercontent.com/47613665/79506553-c6f9cd00-7ffb-11ea-96d4-98eb361cb429.png)
The limited response to the electric field occurs consistently for electric fields up to -1. Even for very, very small electric fields, the response is not symmetric, indicating that the issue is not due to higher-order effects:
![image](https://user-images.githubusercontent.com/47613665/79506697-0fb18600-7ffc-11ea-858c-e7deff3f06ba.png)
Is this an unsupported calculation in PySCF currently, or am I doing something incorrectly? I don't see why this calculation would have these results from a theoretical standpoint, unless I'm missing a crucial detail.

Thank you for your time!

**Comment by sunqm** [comment]
Although the code allows you to call cell.pbc_intor('int1e_r', 3, 1, self.kpt), I think this diple-like integral does not make too much sense with periodic boundary conditions. May be better to compute the integrals on grids like this code did https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/scf/hf.py#L205

**Comment by Simonsays095** [comment]
That function computes a single dipole moment for the whole system, given a density matrix. How would I change it so that I get an output similar to the pbc_intor function, with x-, y-, and z- components of an r x r matrix?

**Comment by sunqm** [comment]
You can compute the matrix elements like https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/scf/hf.py#L253. Instead of using rho there, you can compute the ao-ao product on grids then contracted to r and `grids.weights`. The value of AO can be computed using `cell.pbc_eval_gto('GTOval', grids.coords)`

**Comment by Simonsays095** [comment]
Trying this solution, I think I've gotten much closer. However, there are a few issues I'm having that I'm concerned will affect my results. First of all, I kept things simple by trying to get the overlap matrix first, and later I'll make the 1 small change to get the dipole matrix. Here is my code:
```
import numpy as np
from pyscf.pbc import gto
from pyscf.pbc.dft import gen_grid

cell = gto.Cell()

cell.atom = [
    ['C', (3,3,3)]
]
cell.basis = 'sto-6g'
cell.dimension = 1
cell.a = np.eye(3)*6
cell.unit = 'A'

cell.build()
r = cell.nao_nr()

grids = gen_grid.UniformGrids(cell)
ao = cell.pbc_eval_gto("GTOval", grids.coords)

print("custom overlap matrix")
print(np.einsum('ki,k,kj->ij', ao, grids.weights, ao))

print("overlap matrix from cell directly:")
obc_int = cell.intor("int1e_r", comp=3)
print("molecular:")
print(cell.intor('int1e_ovlp'))

pbc_int = cell.pbc_intor("int1e_r", comp=3)
print("periodic:")
print(cell.pbc_intor('int1e_ovlp'))
```

I'm ignoring E-field in this test code for now. I also changed out the H_4 system for a single carbon atom in a smaller basis for simplicity. I chose a box size of 6 angstroms to avoid as much destructive interference along the periodic axis as possible. Here are the issues I'm having:
 * Using this code, all diagonals are close to 1 (as expected) except for orbital 0, which is around 1.28. This seems like a very large error, since this orbital is no longer normalized when integrating on the grid.
 * The overlap matrix depends heavily on both positioning in the yz (non-periodic) direction, and box size in that direction. For example, changing the C atom to (3,0,0) changes it where orbital 0 has 1 on the diagonal, but the others are near 0.3. If we instead change the box to be `cell.a = np.eye(3)*[6,10,10]`, then orbital 0's diagonal element goes up to 2.42, while the others stay around 1.

I think part of the problem is that changing the box size along non-periodic axes doesn't change the number of grid points in those directions, since the `_mesh_inf_vacuum` function only accounts for `cell.precision` and not `cell.a`. A box 6 angstroms across that has 40 mesh points will also have 40 mesh points if you change it to be 600 angstroms across. I have tested what changing `cell.precision` does, and it does increase the accuracy of the overlap matrix, but even at 1e-15, the diagonal for orbital 0 only converges to 1.02, with a much larger computational cost. With the precision at 1e-15, the calculation with the C atom at (3,0,0) no longer has any eigenvalues near 1.

In summary, here are my main questions:
 * First of all, am I computing the overlap matrix correctly? Assuming I am:
 * How should I choose the position of my atoms in the unit cell to minimize the error I'm seeing?
 * Similarly, how should I choose non-periodic box lengths to minimize this error?

**Comment by gkc1000** [comment]
On Wed, May 20, 2020 at 10:40 AM Simonsays095 <notifications@github.com>
wrote:

> Trying this solution, I think I've gotten much closer. However, there are
> a few issues I'm having that I'm concerned will affect my results. First of
> all, I kept things simple by trying to get the overlap matrix first, and
> later I'll make the 1 small change to get the dipole matrix. Here is my
> code:
>
> import numpy as np
> from pyscf.pbc import gto
> from pyscf.pbc.dft import gen_grid
>
> cell = gto.Cell()
>
> cell.atom = [
>     ['C', (3,3,3)]
> ]
> cell.basis = 'sto-6g'
> cell.dimension = 1
> cell.a = np.eye(3)*6
> cell.unit = 'A'
>
> cell.build()
> r = cell.nao_nr()
>
> grids = gen_grid.UniformGrids(cell)
> ao = cell.pbc_eval_gto("GTOval", grids.coords)
>
> print("custom overlap matrix")
> print(np.einsum('ki,k,kj->ij', ao, grids.weights, ao))
>
> print("overlap matrix from cell directly:")
> obc_int = cell.intor("int1e_r", comp=3)
> print("molecular:")
> print(cell.intor('int1e_ovlp'))
>
> pbc_int = cell.pbc_intor("int1e_r", comp=3)
> print("periodic:")
> print(cell.pbc_intor('int1e_ovlp'))
>
>
One issue is that the uniform grids are inefficient at integrating standard
Gaussian bases, which have some sharp (large exponent) Gaussians in them.
Either using atom-centered grids, or using pseudopotential bases (e.g. GTH)
will fix that problem. Probably easiest to use GTH.

> I'm ignoring E-field in this test code for now. I also changed out the H_4
> system for a single carbon atom in a smaller basis for simplicity. I chose
> a box size of 6 angstroms to avoid as much destructive interference along
> the periodic axis as possible. Here are the issues I'm having:
>
>    - Using this code, all diagonals are close to 1 (as expected) except
>    for orbital 0, which is around 1.28. This seems like a very large error,
>    since this orbital is no longer normalized when integrating on the grid.
>    - The overlap matrix depends heavily on both positioning in the yz
>    (non-periodic) direction, and box size in that direction. For example,
>    changing the C atom to (3,0,0) changes it where orbital 0 has 1 on the
>    diagonal, but the others are near 0.3. If we instead change the box to be cell.a
>    = np.eye(3)*[6,10,10], then orbital 0's diagonal element goes up to
>    2.42, while the others stay around 1.
>
> I think part of the problem is that changing the box size along
> non-periodic axes doesn't change the number of grid points in those
> directions, since the _mesh_inf_vacuum function only accounts for
> cell.precision and not cell.a. A box 6 angstroms across that has 40 mesh
> points will also have 40 mesh points if you change it to be 600 angstroms
> across. I have tested what changing cell.precision does, and it does
>
Someone needs to check this @fishjojo can you take a look?

**Comment by Simonsays095** [comment]
I tried using a gth basis by changing it to `cell.basis='gthdzv'`, but the results are not much better:

**Box: 6 A Cube
System: A single carbon atom**

| Basis | Atom Position | Precision | Diagonals  | | | | |
| ----- |  --------------- | --------- | ------------- | --- | --- | --- | --- |
| sto6g | (3,3,3) | 1e-8 | 1.284 | 1 | 1 | 1 | 1 |
| sto6g | (3,3,3) | 1e-20 | 1.007 | 1 | 1 | 1 | 1 |
| sto6g | (0,0,0) | 1e-8 | 1.019 | 0.314 | 0.351 | 0.296 | 0.296 |
| sto6g | (0,0,0) | 1e-20 | 0.439 | 0.274 | 0.287 | 0.268 | 0.268 |
| gthdzv | (3,3,3) | 1e-8 | 1 | 1.001 | 0.998 | 1 | 1 | 0.986 | 0.999 | 0.999 |
| gthdzv | (3,3,3) | 1e-20 | 1 | 1.001 | 0.998 | 1 | 1 | 0.986 | 0.999 | 0.999 |
| gthdzv | (0,0,0) | 1e-8 | 0.306 | 0.291 | 0.339 | 0.291 | 0.291 | 0.286 | 0.270 | 0.270 |
| gthdzv | (0,0,0) | 1e-20 | 0.271 | 0.265 | 0.282 | 0.266 | 0.266 | 0.262 | 0.258 | 0.258 |

When the carbon is at the center of the cell, it makes the results better, but that doesn't fix the translational variance. When the carbon atom is at the origin, results are always far from 1.

Can you confirm that the results of `np.einsum('ki,k,kj->ij', ao, grids.weights, ao)` should be an overlap matrix, so the diagonals should be 1? If that's the case, then either I'm missing another source of error, there's a bug in the `pbc_eval_gto` function, or atom positioning guidelines for unit cells needs to be clarified.

**Comment by gkc1000** [comment]
Is cell dimensionality = 3 or 1?

**Comment by Simonsays095** [comment]
The cell dimension is 1.

**Comment by gkc1000** [comment]
The integral should be the overlap matrix.

I think as mentioned in your previous issue, what is happening is the way
the grid is set up in 1D does not use the precision parameter along the
non-periodic axes. This is something needs to be checked @fishjojo. So even
though the GTH basis functions are easy to integrate, you are still getting
a large error.

However, there are several ways around this:
1. Use cell.dim = 3 (if you are worried about the periodic images you can
always increase the vacuum size in the two directions later).
2. What is the bigger picture? If you are treating the dipole coupling to
an electric field, in a periodic system you can always choose the gauge so
that the coupling becomes p.A (this is what is done in most periodic
calculations  (see e.g.  https://arxiv.org/pdf/1907.11695.pdf)).

**Comment by Simonsays095** [comment]
The big picture is that I want to probe the energetic response to an applied electric field, to get an idea of the polarizability of the system (per unit cell). The single carbon atom I have been using in the tests is just a simple model system to make sure I'm doing it right. I can move the gauge as the code @sunqm described in an earlier comment, but that only takes effect in the dipole moment integral - this is an issue with the AO grid itself, which is before any gauge can be used.

I can try the `cell.dim = 3` solution, but some quick testing with the C atom indicates that I'm going to need a huge box to remove the periodic effects in the extra 2 periodic dimensions. With box lengths of [6, 100, 100], the energy of the 3-D calculation still differs from the 1-D energy by ~1 mH, despite taking far more computation/memory to compute the grid points for the far larger box.

**Comment by gkc1000** [comment]
When you were examining the finite size effects in 3D, were you doing a HF
or a pure DFT calculation?

Re: improving the grid integration in 1D, there are the following options
...
1. Use the Becke grids (it looks like you are using the evenly spaced
grids). This would allow you to integrate any Gaussian accurately,
2. Set up the evenly spaced grid coords yourself in the 1D box, one just
needs to set up the x, y, z coords

**Comment by Simonsays095** [comment]
For the finite size effects, I did a simple HF calculation. I don't plan on using DFT in the long run, I'm going to use a periodic version of the v2RDM method I've been making.

I think the problem I'm having here is actually worse than just the mesh density in 1-dimension. To test it, I set up both a 1-D and a 3-D calculation with a box size of 6x50x50 angstroms, with precision set to 1e-8 (default). I commented out the [3 lines of code](https://github.com/pyscf/pyscf/blob/7ba318ff7cae650b255d5c5b90ed550521d324f0/pyscf/pbc/gto/cell.py#L1373-L1375) that overwrite the non-periodic mesh values with the `_mesh_inf_vacuum` value, and verified that the mesh was the same for each system, at [82, 682, 682]. Here are the results:

Dimensions | Energy | Overlap Diagonals |  |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | --
3-D | -14.7934 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001
1-D | -14.7933 | 0.277 | 0.270 | 0.291 | 0.270 | 0.270 | 0.266 | 0.260 | 0.260

This seems to suggest that there's another problem, in addition to the mesh density.

**Comment by gkc1000** [comment]
That is odd. Well, this actually requires looking into, I don't have time
to look into it myself. I'll have to tag help wanted. @Xiao Wang
<xwang@flatironinstitute.org> do you have time to look into this?

**Comment by fishjojo** [comment]
> For the finite size effects, I did a simple HF calculation. I don't plan on using DFT in the long run, I'm going to use a periodic version of the v2RDM method I've been making.
> 
> I think the problem I'm having here is actually worse than just the mesh density in 1-dimension. To test it, I set up both a 1-D and a 3-D calculation with a box size of 6x50x50 angstroms, with precision set to 1e-8 (default). I commented out the [3 lines of code](https://github.com/pyscf/pyscf/blob/7ba318ff7cae650b255d5c5b90ed550521d324f0/pyscf/pbc/gto/cell.py#L1373-L1375) that overwrite the non-periodic mesh values with the `_mesh_inf_vacuum` value, and verified that the mesh was the same for each system, at [82, 682, 682]. Here are the results:
> 
> Dimensions	Energy	Overlap Diagonals		 	 	 	 	 
> 3-D	-14.7934	1.000	1.001	0.998	1.000	1.000	0.986	1.001
> 1-D	-14.7933	0.277	0.270	0.291	0.270	0.270	0.266	0.260
> This seems to suggest that there's another problem, in addition to the mesh density.

How did you compute the overlap matrix.
I got same results for 1D and 3D even with a box size of 6x15x15.
Why is your mesh so small? For the carbon atom system you provided, I got a mesh size of 825x40x40 for 1D and 825x2062x2062 for 3D.

**Comment by Simonsays095** [comment]
I used the `gthdzv` basis for this specific calculation, sorry. Apparently for this system, it has a much smaller grid. I tried doing the calculation with sto6g, but by computer can't handle the memory requirements for 825x2062x2062 grid points.

As for the results I cited above:
 * I computed the overlap matrix as I described earlier in the thread, using `pbc_eval_gto`, so that I can extend it to other more complicated integrals (namely, the dipole integral matrix) once I understand this simple one.
 * I think I accidentally compared a 3-D calculation with the C at (3,3,3) to a 1-D calculation with the carbon at the origin. Double-checking my code and making sure I'm comparing apples to apples, it seems that just changing the dimension does, in fact, leave the results the same (if the mesh is not changed, and the box is large). However, atom positioning still plays a huge role in the outcome, regardless of dimensionality. I also did the calculation setting `cell.precision = 1e-50`, but the diagonals actually got worse by ~0.01 each for the origin-centered atom.
 * Here is the full code snippet:
```
import numpy as np
from pyscf.pbc import gto, scf
from pyscf.pbc.dft import gen_grid

def ovlp_mat(cell):
    grids = gen_grid.UniformGrids(cell)
    ao = cell.eval_gto("GTOval", grids.coords)

    return np.einsum('ki,k,kj->ij', ao, grids.weights, ao)

cell = gto.Cell()

cell.atom = [
    ['C', (0,0,0)] # Change position to (3,3,3) here
]

cell.basis = 'gthdzv'
cell.dimension = 1 # Change to 3 for other calculation
cell.a = np.eye(3)*[6, 50, 50]
cell.unit = 'A'
cell.build()

ovlp = ovlp_mat(cell)
diag = ovlp.diagonal()

print("Calculating energy")
mf = scf.RHF(cell).density_fit()
mf.kernel() # gamma-point calculation
E_HF = mf.e_tot

print("Energy\t\t\tDiagonals")
print(E_HF, *diag, sep="\t")
```
And here are the results:

dimension | C position | Energy | overlap   diagonals |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1-D | (0,0,0) | -14.7933 | 0.145641 | 0.139828 | 0.145895 | 0.145873 | 0.145873 | 0.134706 | 0.134696 | 0.134696
3-D | (0,0,0) | -14.7934 | 0.145641 | 0.139828 | 0.145895 | 0.145873 | 0.145873 | 0.134706 | 0.134696 | 0.134696
1-D | (3,3,3) | -14.7933 | 0.999973 | 0.99983 | 0.999722 | 0.999868 | 0.999868 | 0.99831 | 0.999201 | 0.999201
3-D | (3,3,3) | -14.7934 | 0.999973 | 0.99983 | 0.999722 | 0.999868 | 0.999868 | 0.99831 | 0.999201 | 0.999201

**Comment by gkc1000** [comment]
I think what is happening is in calling cell.eval_gto() it is passing to
mole.eval_gto(), because to get the PBC version of it you have to have pbc
in the first argument, see

https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/gto/cell.py

so, without the periodic images, the overlap matrix will be very different
where you place the atom.

**Comment by gkc1000** [comment]
Also, if you don't have enough memory for using a uniform grid, then use gen_grid.get_becke_grids instead, they should be a lot smaller.

**Comment by Simonsays095** [comment]
You're right, that was a mistake that I left in after trying to compare this manual overlap calculation method between open and periodic boundary conditions. After changing that back to `pbc_eval_gto`, the position and dimensionality no longer changes the diagonal elements - but only when using the Becke grids instead of uniform grids.
 * Using the sto6g has all diagonals within within 5e-5 of 1, whereas gthdzv has one diagonal element 0.014 away, and another 0.0025 away.
 * The mesh density is still unaffected by precision in the non-periodic axes, but this somehow doesn't affect the overlap matrix at all when using Becke grids. I increased the box size to 6x1e5x1e5 and the mesh was still 825x40x40, yet all diagonal elements were unchanged.
 * Now that it looks like I got the overlap matrix working (using Becke grids for accuracy) I used this method to create the dipole matrix, and it seems to work properly (although I need to do more testing to be sure). The y- and z- components are identical to `pbc_intor('int1e_r'. comp=3)` whereas the x- component is changed. This is the component I was having problems with in 1-D, so I'm hopeful that this solves my problem.

**Comment by gkc1000** [comment]
1. When you say the diagonal elements are 0.014 away and 0.0025 away, can you confirm if that is for the Becke grid or the small mesh that you are using?

2. The Becke grid level is independent of the mesh density. The mesh density is associated with FFTDF and for the smooth part of the Coulomb potential in GDF, i.e. it is used for the Coulomb problem. One-electron integrals in PBC are either done analytically via lattice summation or using Becke grids (for vxc).

3. In the bigger picture, to do what you want to do, you need to address the issue of dipole origin. If you are just going to fix it at the charge center of the cell, then you could use pbc_intor anyways, and replace the lattice summation in pbc_intor by a 1D lattice summation. Otherwise, it is more typical to represent the position operator by d/dG in momentum space for PBC.

**Comment by xwang862** [comment]
>I think I accidentally compared a 3-D calculation with the C at (3,3,3) to a 1-D calculation with the carbon at the origin. Double-checking my code and making sure I'm comparing apples to apples, it seems that just changing the dimension does, in fact, leave the results the same (if the mesh is not changed, and the box is large). However, atom positioning still plays a huge role in the outcome, regardless of dimensionality. I also did the calculation setting cell.precision = 1e-50, but the diagonals actually got worse by ~0.01 each for the origin-centered atom.
>
>Here is the full code snippet:
>```import numpy as np
>from pyscf.pbc import gto, scf
>from pyscf.pbc.dft import gen_grid
>
>def ovlp_mat(cell):
>    grids = gen_grid.UniformGrids(cell)
>    ao = cell.eval_gto("GTOval", grids.coords)
>
>    return np.einsum('ki,k,kj->ij', ao, grids.weights, ao)
>
>cell = gto.Cell()
>
>cell.atom = [
>    ['C', (0,0,0)] # Change position to (3,3,3) here
>]
>
>cell.basis = 'gthdzv'
>cell.dimension = 1 # Change to 3 for other calculation
>cell.a = np.eye(3)*[6, 50, 50]
>cell.unit = 'A'
>cell.build()
>
>ovlp = ovlp_mat(cell)
>diag = ovlp.diagonal()
>
>print("Calculating energy")
>mf = scf.RHF(cell).density_fit()
>mf.kernel() # gamma-point calculation
>E_HF = mf.e_tot
>
>print("Energy\t\t\tDiagonals")
>print(E_HF, *diag, sep="\t")
>```

I ran the above script with one modification - replacing `ao = cell.eval_gto("GTOval", grids.coords)` with `cell.pbc_eval_gto("GTOval", grids.coords)`, as @gkc1000 pointed out. Also I commented out [the three lines of code](https://github.com/pyscf/pyscf/blob/7ba318ff7cae650b255d5c5b90ed550521d324f0/pyscf/pbc/gto/cell.py#L1373-L1375). My results below got reproduce @Simonsays095's energies but not overlap diagonals (probably because I used the default `cell.precision`). 

dimension | C position | Energy | overlap diagonals |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1-D | (0, 0, 0) | -14.79326859 | 0.277 | 0.270 | 0.291 | 0.270 | 0.270 | 0.266 | 0.260 | 0.260
3-D | (0, 0, 0) | -14.79341573 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001
1-D | (3, 3, 3) | -14.79326859 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.000 | 1.000
3-D | (3, 3, 3) | -14.79341573 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001

Anyway, only the 1-D results show remarkably different overlap diagonals, depending on the position of C. The 3-D case is independent of atom positions. @Simonsays095 can you please double check your 3-D calculation with C at (0,0,0)?

**Comment by xwang862** [comment]
As Garnet suggested, using the Becke grid is probably a better idea (
`grids = gen_grid.BeckeGrids(cell); grids.build()`). The following results computed with BeckeGrids give correct diagonals for 1-D at either atom position:

dimension | C position | Energy | overlap diagonals |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1-D | (0, 0, 0) | -14.79326859 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001
3-D | (0, 0, 0) | -14.79341573 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001
1-D | (3, 3, 3) | -14.79326859 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001
3-D | (3, 3, 3) | -14.79341573 | 1.000 | 1.001 | 0.998 | 1.000 | 1.000 | 0.986 | 1.001 | 1.001

Btw, another way to compute PBC 1-e integrals is through [`pbc_intor`](https://github.com/pyscf/pyscf/blob/7ba318ff7cae650b255d5c5b90ed550521d324f0/pyscf/pbc/gto/cell.py#L1607). E.g. the overlap matrix can be easily obtained by `mf.get_ovlp(cell)`, which is just a wrapper to `pbc_intor`.

**Comment by xwang862** [comment]
Oh sorry I missed your latest comments @Simonsays095. Seems like you already figured out most of the issue 👍

**Comment by Simonsays095** [comment]
> 1. When you say the diagonal elements are 0.014 away and 0.0025 away, can you confirm if that is for the Becke grid or the small mesh that you are using?

As you can see in @xwang862's comments, it comes from a gthdzv basis with a Becke grid.

> 2. The Becke grid level is independent of the mesh density. The mesh density is associated with FFTDF and for the smooth part of the Coulomb potential in GDF, i.e. it is used for the Coulomb problem. One-electron integrals in PBC are either done analytically via lattice summation or using Becke grids (for vxc).

> 3. In the bigger picture, to do what you want to do, you need to address the issue of dipole origin. If you are just going to fix it at the charge center of the cell, then you could use pbc_intor anyways, and replace the lattice summation in pbc_intor by a 1D lattice summation. Otherwise, it is more typical to represent the position operator by d/dG in momentum space for PBC.

I understand that I will need to set a dipole origin, and my plan is to use the charge center as you point out. However, according to @sunqm's comments at the very beginning of this thread, he believes that `cell.pbc_intor('int1e_r', 3, 1, self.kpt)` will give erroneous results for some reason. That's why I started computing the integrals on a grid in the first place. Comparing the dipole integrals I compute on the grid to the `pbc_intor` value, it seems like the components for non-periodic axes are identical, but the component for periodic axes are indeed different.

**Comment by fishjojo** [comment]
The uniform grid is always centered in the cell. So if you put the atom at the edge of the cell, there won't be enough grid points in the non-periodic directions. To get the best result, you should put the atom in the center of the cell. Apparently, atomic grids don't have this problem.

**Comment by Simonsays095** [comment]
As I mentioned in an earlier reply, that's one of the things I think could be better documented, because the actual position of the cell in space is not documented anywhere that I could find. I believe (after looking into the code and running tests) that the unit cell is located in the positive octet with one vertex of the parallelopiped at the origin, and lattice vectors defined by `cell.a`. But this certainly can't always be true, since some unit cell shapes will either require some of the cell to be in a different octet, or for the vertex to leave the origin.

For example, where in space would a unit cell with the following `a` matrix be located?

1 | 0 | 0
-- | -- | --
0 | 1 | -1
0 | 0 | -1

**Comment by gkc1000** [comment]
The unit cell vectors define the parallelpiped starting at 0; this is the
usual definition of a unit cell.
The mesh is defined as fractional coordinates along the unit cell vectors.
But I think this is off-topic, I believe the original query about how to
compute a dipole moment in a non-periodic cell using a grid has been
answered. If there are no more questions on that issue, I will close it.
